### PR TITLE
Add Browser test configuration and scripts

### LINKED ISSUE #6: Tests for browser
datacook aims to work for both browser and nodejs. We should construct unit test for both to ensure consistency.

### PR DESCRIPTION
**Pull request by risenW**
This pull fixes #6 : 
* Adds Karma as a browser testing library
* Add Webpack bundler for generating browser build
* Add first browser test for Generic class

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,4 +6,5 @@ danfojs/data
 .DS_Store
 .idea/
 dist/
-tsconfig.tsbuildinfo
\ No newline at end of file
+tsconfig.tsbuildinfo
+index.html
```
Comment (WenheLI) [inline .gitignore:10]: EOF

**Comment by WenheLI** [comment]
Will this PR also cover image ops' test?

**Comment by risenW** [comment]
> Will this PR also cover image ops' test?

Yes it will, I have not added it though. We have to write the same tests with just some minimal changes. most especially in imports.

File: test/browser-test/generic/index.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+const tf = require("@tensorflow/tfjs-core");
```
Comment (yorkie) [inline test/browser-test/generic/index.js:1]: Shall we use ES Modules?

File: webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+const path = require('path');
+
+const createConfig = (target) => {
+  return {
+    mode: 'production',
```
Comment (yorkie) [inline webpack.config.js:5]: I think we should keep a consistent style that we used double quotes before.

**Comment by yorkie** [comment]
@risenW Where do we run tests? How about https://www.browserling.com/ :)

File: test/browser-test/generic/index.js (hunk 2 above)
Comment (risenW) [reply test/browser-test/generic/index.js:1]: True, I'll fix that

**Comment by risenW** [comment]
> @risenW Where do we run tests? How about https://www.browserling.com/ :)

Tests are run from the terminal as well, but Karma runs this in the [specified browser](https://karma-runner.github.io/5.2/config/browsers.html) (Chrome currently).

**Comment by yorkie** [comment]
Oh, visual testing might not be required to use it, however, it provides multiple environments(browser-side) for running our JavaScript, not sure if the webdirver is working there.

@risenW never mind, now the Karma's headless environment looks good to me :)

File: test/browser-test/generic/index.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+describe("Generic Split test in browser", () => {
+  it("should split data into train & test", async () => {
+ 
+    const X = tf.range(0, 10).reshape([5, 2]);
+
+    const y = tf.range(0, 5);
+
+    const [X_train, X_test, y_train, y_test] = datacook.Generic.split([X, y]);
```
Comment (yorkie) [inline test/browser-test/generic/index.js:10]: Why does the X to be uppercase?

File: test/node/image/image-proc.ts [hunk 5]
```diff
@@ -1,64 +1,64 @@
 import { assert, expect } from "chai"
-import Image from "../../src/image/image-proc";
-import {stdCalc} from "../../src/image/utils";
+import Image from "../../../src/image/image-proc";
+import {stdCalc} from "../../../src/image/utils";
```
Comment (yorkie) [inline test/node/image/image-proc.ts:3]: This relative path looks too long, shall we have an alias config for the directory "../../../src" as "@/image/utils"?

File: README.md [hunk 6]
```diff
@@ -12,21 +12,30 @@ cd datacook
...
+npm run test:node
 ```
 
-or if you use npm
+or for browser:
+```
+npm run test:browser
+```
+
+To build/bundle the source code to the dist folder, run:
```
Comment (yorkie) [inline README.md:35]:
```suggestion
To build or bundle the source code to the dist folder, run:
```

File: test/browser-test/generic/index.js (hunk 4 above)
Comment (WenheLI) [reply test/browser-test/generic/index.js:10]: this aligns to python convention

File: webpack.config.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    devtool: 'source-map',
+    context: path.resolve(__dirname),
+    entry: {
+      index: `./dist/index.js`
```
Comment (FeelyChau) [inline webpack.config.js:9]:
```suggestion
      index: './dist/index.js'
```

File: webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    },
+    target: target,
+    output: {
+      path: path.resolve(__dirname, 'dist/browser'),
+      filename: `bundle.js`,
```
Comment (FeelyChau) [inline webpack.config.js:14]: Ditto.

File: karma.conf.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
+
+module.exports = function (config) {
+  config.set({
+    // base path that will be used to resolve all patterns (eg. files, exclude)
+    basePath: "",
+
+    // frameworks to use
+    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
+    frameworks: [ "jasmine", "browserify" ],
```
Comment (FeelyChau) [inline karma.conf.js:9]: Should we import `jasmine` when `mocha` has been used?

**Review by FeelyChau — DISMISSED** [review]

File: test/browser-test/generic/index.js (hunk 2 above)
Comment (risenW) [reply test/browser-test/generic/index.js:1]: I just found out that Karma does not support ES6 import from installed packages, that means I cannot import an installed library in-browser test file using ES6 format. You can only import relative scripts. 

Ref https://stackoverflow.com/questions/57066106/typeerror-failed-to-resolve-module-specifier-open-wc-testing-relative-refer didn't help either. 

So we can stick to using require for browser testing unless there is a better solution

File: test/browser-test/generic/index.js (hunk 4 above)
Comment (yorkie) [reply test/browser-test/generic/index.js:10]: Hmm, why do we keep python convention on a JS project?

File: test/browser-test/generic/index.js (hunk 2 above)
Comment (yorkie) [reply test/browser-test/generic/index.js:1]: How about using webpack `resolve.alias`? See https://webpack.js.org/configuration/resolve/.

**Comment by WenheLI** [comment]
@risenW - Gentle ping, any progress on this thread

File: test/browser-test/generic/index.js (hunk 2 above)
Comment (risenW) [reply test/browser-test/generic/index.js:1]: @yorkie This more of a Karma issue not webpack

File: karma.conf.js (hunk 9 above)
Comment (risenW) [reply karma.conf.js]: Fixed, thanks for the headsup

**Comment by risenW** [comment]
> @risenW - Gentle ping, any progress on this thread

I have fixed most issues raised here. Ready for review

**Comment by WenheLI** [comment]
@risenW - Thanks for the amazing work! Most of the code looks good to me, just one more thing, we should remove the yarn-error.log from the commit.

**Comment by risenW** [comment]
> @risenW - Thanks for the amazing work! Most of the code looks good to me, just one more thing, we should remove the yarn-error.log from the commit.

Done

**Review by WenheLI — APPROVED** [review]
LGTM

**Pull request merged**